### ISSUE TITLE
[OpenCL] Compile error: casting changes address space of pointer

### ISSUE THREAD

**Issue by eltos** [xsuite/xsuite#247]
I am trying to get xsuite running on an AMD GPU with the OpenCL context using the [simple example](https://xsuite.readthedocs.io/en/latest/singlepart.html#a-simple-example) from the getting started guide. I get the following error:
```
Traceback (most recent call last):
  File "/xsuite/example.py", line 32, in <module>
    tracker = xt.Tracker(_context=context, line=line)
  File "/xsuite/xtrack/xtrack/tracker.py", line 101, in __init__
    self._init_track_no_collective(
  File "/xsuite/xtrack/xtrack/tracker.py", line 351, in _init_track_no_collective
    self._build_kernel(save_source_as, compile=compile)
  File "/xsuite/xtrack/xtrack/tracker.py", line 722, in _build_kernel
    context.add_kernels(
  File "/xsuite/xobjects/xobjects/context_pyopencl.py", line 271, in add_kernels
    prg = cl.Program(self.context, specialized_source).build()
  File "/usr/local/lib/python3.10/dist-packages/pyopencl/__init__.py", line 535, in build
    self._prg, was_cached = self._build_and_catch_errors(
  File "/usr/local/lib/python3.10/dist-packages/pyopencl/__init__.py", line 583, in _build_and_catch_errors
    raise err
pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'gfx906:sramecc+:xnack-' on 'AMD Accelerated Parallel Processing' at 0x564a3eb54d60>:

/tmp/comgr-f5172f/input/CompileSource:244:10: error: casting '__global char *' to type '__private char *' changes address space of pointer
  return (char*)(( __global char*) obj+offset);
         ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-f5172f/input/CompileSource:4612:10: error: casting '__global char *' to type '__private char *' changes address space of pointer
  return (char*)(( __global char*) obj+offset);
         ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-f5172f/input/CompileSource:4672:31: error: initializing '__private void *__private' with an expression of type '__global void *' changes address space of pointer
                        void* el = ElementRefData_member_elements(elem_ref_data, elem_idx);
                              ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-f5172f/input/CompileSource:4687:56: error: casting '__private void *' to type 'DriftData' (aka '__global struct DriftData_s *') changes address space of pointer
                            Drift_track_local_particle((DriftData) el, &lpart);
                                                       ^           ~~
/tmp/comgr-f5172f/input/CompileSource:4693:60: error: casting '__private void *' to type 'MultipoleData' (aka '__global struct MultipoleData_s *') changes address space of pointer
                            Multipole_track_local_particle((MultipoleData) el, &lpart);
                                                           ^               ~~
/tmp/comgr-f5172f/input/CompileSource:4699:67: error: casting '__private void *' to type 'ParticlesMonitorData' (aka '__global struct ParticlesMonitorData_s *') changes address space of pointer
                            ParticlesMonitor_track_local_particle((ParticlesMonitorData) el, &lpart);
                                                                  ^                      ~~
6 errors generated.
Error: Failed to compile source (from CL or HIP source to LLVM IR).

(options: -I /usr/local/lib/python3.10/dist-packages/pyopencl/cl)
(source saved as /tmp/tmp3o3ucscd.cl)
```
[tmp3o3ucscd.cl.txt](https://github.com/xsuite/xobjects/files/9926398/tmp3o3ucscd.cl.txt)

My setup is an Ubuntu 22.04 with AMD OpenCL drivers (amdgpu-install --usecase=opencl), python 3.10 and the latest xsuite modules (git clone and pip install -e with the fix described in xsuite/xobjects#68 applied):
```
numpy               1.21.5
cffi                1.15.1
Cython              0.29.32
xdeps               0.0.10
xfields             0.8.4
xobjects            0.1.26
xpart               0.9.2
xsuite              0.4.0
xtrack              0.21.0
pyopencl            2022.2.4
```

If I select the CPU as device in the opencl context (`<pyopencl.Device 'pthread-12th Gen Intel(R) Core(TM) i5-1235U' on 'Portable Computing Language' at 0x5559ba82e500>`) I get the same error message.

**Comment by szymonlopaciuk** [comment]
In order to resolve this problem for now I suggest indeed building against OpenCL v2 which should be accomplished by https://github.com/xsuite/xobjects/pull/80. I will test the PR on an AMD machine myself, but please let me know if you have any issues with this.

In principle it should be possible to make this work with CL 1.2, but I am not sure if there is a need. What do you think?

**Comment by rdemaria** [comment]
It looks like a bug in the C side. A change of the address space does not seem intentional to me. In addition, OpenCL 2.0 is not supported by many vendors.

**Comment by giadarol** [comment]
Indeed it was a bug in the tracker source (tolerated by cuda and OpenCL 2.0)
Pull request https://github.com/xsuite/xtrack/pull/263 should fix the bug and restore compatibility with OpenCL 1.3

**Comment by rdemaria** [comment]
Compare the two outputs for <class>_getp1:
```
print(xo.Float64[:]._gen_c_api())
print(xo.String[:]._gen_c_api())
```
`/*gpuglmem*/` is missing in a few places. Probably also 4 errors.

**Comment by rdemaria** [comment]
There is also a bug in `gen_method_getp`. It should be `if is_scalar(lasttype) or is_string(lasttype)`. But I need to add a test.

**Comment by giadarol** [comment]
I agree.

Pull request https://github.com/xsuite/xtrack/pull/263 should solve these ones:
```
/tmp/comgr-f5172f/input/CompileSource:4672:31: error: initializing '__private void *__private' with an expression of type '__global void *' changes address space of pointer
                        void* el = ElementRefData_member_elements(elem_ref_data, elem_idx);
                              ^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-f5172f/input/CompileSource:4687:56: error: casting '__private void *' to type 'DriftData' (aka '__global struct DriftData_s *') changes address space of pointer
                            Drift_track_local_particle((DriftData) el, &lpart);
                                                       ^           ~~
/tmp/comgr-f5172f/input/CompileSource:4693:60: error: casting '__private void *' to type 'MultipoleData' (aka '__global struct MultipoleData_s *') changes address space of pointer
                            Multipole_track_local_particle((MultipoleData) el, &lpart);
                                                           ^               ~~
/tmp/comgr-f5172f/input/CompileSource:4699:67: error: casting '__private void *' to type 'ParticlesMonitorData' (aka '__global struct ParticlesMonitorData_s *') changes address space of pointer
                            ParticlesMonitor_track_local_particle((ParticlesMonitorData) el, &lpart);
                                                                  ^                      ~~
```

While these ones come from xobjects:
```
/tmp/comgr-f5172f/input/CompileSource:244:10: error: casting '__global char *' to type '__private char *' changes address space of pointer
  return (char*)(( __global char*) obj+offset);
         ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/comgr-f5172f/input/CompileSource:4612:10: error: casting '__global char *' to type '__private char *' changes address space of pointer
  return (char*)(( __global char*) obj+offset);
         ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by giadarol** [comment]
We should probably find a way of testing regularly on AMD

**Comment by rdemaria** [comment]
The problem was not only amd, I could reproduce on intel opencl as well!

**Comment by rdemaria** [comment]
I found a fix in #82. The PR is messy because there are many changes on the pipeline for xobjects, it is two lines.
I added also a test in test_capi that tests GPUs and that can be extended to test part of the C Api on gpus.

The fix was particularly tricky. Initially, I thought to return `char **` for `ArrayNString_getp1` which explicitly should become `__global char * __global *` , then I opted for a simpler `_global char *`
